test/system_metrics_e2e_test.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+End to end tests for the system_metrics_collector ROS2 package.
+
+These tests launch the talker_listener_example and use ROS2 to inspect if the
+expected nodes are active and publishing data. To run these tests ensure that
+ROS2 is installed, with the system_metrics_collector package the relevant
+setup.bash has been sourced.
+"""
+
+import logging
+import signal
+import subprocess
+import sys
+
+import test_helpers
+
+import rclpy
+
+# expected outputs
+EXPECTED_LIFECYCLE_NODES = frozenset(['/linux_system_cpu_collector',
+                                      '/linux_system_memory_collector',
+                                      '/listener_process_cpu_node',
+                                      '/listener_process_memory_node',
+                                      '/talker_process_cpu_node',
+                                      '/talker_process_memory_node'])
+EXPECTED_REGULAR_NODES = frozenset(['/listener', '/talker'])
+EXPECTED_LIFECYCLE_STATE = 'active [3]'
+EXPECTED_TOPIC = '/system_metrics'
+# executable commands
+LAUNCH_COMMAND = 'ros2 launch system_metrics_collector talker_listener_example.launch.py'
+# test constants
+TIMEOUT_SECONDS = 30
+RETURN_VALUE_FAILURE = 1
+RETURN_VALUE_SUCCESS = 0
+EXPECTED_NUMBER_OF_MESSAGES_TO_RECEIVE = 5
+
+def setup_logger() -> None:
+    """Format and setup the logger."""
+    logger = logging.getLogger()
+    # setting to debug for end to end test soak
+    logger.setLevel(logging.DEBUG)
+
+    handler = logging.StreamHandler(sys.stdout)
+    handler.setLevel(logging.DEBUG)
+    formatter = logging.Formatter('[%(module)s] [%(levelname)s] [%(asctime)s]: %(message)s')
+    handler.setFormatter(formatter)
+    logger.addHandler(handler)
+
+
+def main(args=None) -> int:
+    """
+    Run all the e2e tests. This exits on the first failure encountered.
+
+    :param args:
+    :return: 0 if all tests pass, 1 if any fail
+    """
+
+    try:
+        rclpy.init()
```

Comment:
Why init here? This is only needed for the statistics listener test, as everything else should be self contained.

Reply:
Calling `rclpy.init()` from anywhere but `main()` caused `NotInitializedException` to be thrown while creating the `StatisticsListener` node.